### ISSUE TITLE
systemd: open files containing secrets for services

### ISSUE THREAD

**Issue by aburch** [systemd/systemd#14264]
**Is your feature request related to a problem? Please describe.**
Many services should run as an unprivileged user, but need access to some secret data containing certificates (or more exact: secret keys for asymmetric cryptography), passwords, or similar.
Currently you either put these into files that are readable by the service user (also after the service has initialized) or start the service as `root` and require the service to drop its privileges itself.

This was similar to binding to network ports <1024, but systemd allowed services to have systemd open the socket for them instead.

I was wondering if systemd could provide similar functionality for configuration files containing private data.

**Describe the solution you'd like**
.service units could specify configuration files that the systemd manager would open (readonly) when starting the service and pass the open file descriptor to the service, just like it can already do for network sockets.

The service should be informed of this via environment variables (that give file descriptors and associated names to identify different files similar to `LISTEN_FDS` and `LISTEN_FD_NAMES`)

**Comment by boucman** [comment]
maybe some way for an ExecStartPre process to open the file then store it in LISTEN_FDS ?

Since ExecStartPre can be run as root even when User= is used, that should be good enough...

**Comment by aburch** [comment]
Boucman wrote:
> maybe some way for an ExecStartPre process to open the file then
> store it in LISTEN_FDS ?
>
> Since ExecStartPre can be run as root even when User= is used, that
> should be good enough...

Hmm, I guess opening a file in `ExecStartPre` and passing it to systemd
with sd_notify's FDSTORE=1 FDNAME=... might technically work.

But a declarative way to do so might be nicer.  Something like
```
Open=/etc/my/secret/key.pem
```
(though `Open` might be a bit too short a name) which would use
`FDNAME=key.pem` (maybe with some way to specify the name).

A daemon could then allow to specify
```
tls_private_key = systemd:key.pem
```
instead of a file.

As a possible future extension one could also allow opening file in
read-write mode if there are useful applications to do so.

**Comment by zimbatm** [comment]
A few questions that come to mind:

What should happen if the secret gets rotated. Is it the responsibility of systemd to monitor and reload the service? How does the service get access to the new secret? If the secret is stored in a remote location (eg: HashiCorp Vault), how would that be handled?

**Comment by poettering** [comment]
I must say I like @aburch's idea of an Open= a lot. Seems, simple and generic.

**Comment by poettering** [comment]
(this should be modelled vaguely after StandardInput='s syntax I figure)

(also, it might be worth thinking about relationship with the kernel keyring here... i.e. maybe we should also have KeyRingLoad= or so, which uploads a file to the kernel keyring. not saying this would mean we shouldn't implement Open=, but I think it's something to at least think a bit abou

**Comment by poettering** [comment]
> What should happen if the secret gets rotated. Is it the responsibility of systemd to monitor and reload the service? How does the service get access to the new secret? If the secret is stored in a remote location (eg: HashiCorp Vault), how would that be handled?

Like for all configuration changes we expect the user to care about what to do when config changes...

**Comment by aburch** [comment]
zimbatm writes:
> What should happen if the secret gets rotated. Is it the
> responsibility of systemd to monitor and reload the service?

No, the user is responsible just as for other config changes.

> How does the service get access to the new secret?

Through restart (not reload) and getting a new file descriptor.

This is a trade off. Alternatively services could reload:

- The service could keep the file open, but this doesn't work if the
  secret is in a new file, including when the new file has the same name
  (i.e. new file that replaces the old one). I believe this might be
  confusing.

- Implement some other mechanism to request file descriptors, but this
  is more complicated and might require less strict sandboxing for some
  configurations (to talk to whatever would distribute the secret).

If a service cannot be restarted (only reloaded) for some reason, it
should be no problem to just not use this feature.

> If the secret is stored in a remote location (eg: HashiCorp Vault),
> how would that be handled?

The secret could be written to some local file (on a tmpfs) and the
depending service started one it is available. But I believe that is
outside the part systemd needs to deal with.

If a service talks directly to an external secret store, don't use the
proposed feature.

**Comment by cedws** [comment]
:+1: I'd be interested in something like this, but it's also possible to use `EnvironmentFile` as a workaround if your application allows you to pass these secrets as environment variables. Your `EnvironmentFile` can be owned by `root` w/ `600` and still pass secrets to an unprivileged service.

[Shameless plug of my blog post which might describe this better](https://cedwards.xyz/protecting-systemd-service-secrets/)

**Comment by keszybz** [comment]
Implemented in v247: bb0c0d6f29.
